File: lib/pipeline/hue-reserve.ts
import { oklabDistanceSquared, oklabToRgb, rgbToOklab, type Oklab } from "../color/color";
import { EMPTY_CELL, type RGB } from "../types";

/**
 * Vivid, part two (G-062, D212): a thread for a hue the photo has, whether or not it earns one by area.
 *
 * k-means allocates by squared error, so a colour covering half a percent of the chart loses to one more step in the
 * dominant ramp however visible it is — measured across G-060 and G-061, the cat photo's pink got no thread at all by
 * 64 colours under four different mechanisms. This one does not ask it to win that contest: each hue the cells hold
 * and the palette does not speak for takes a slot, paid for by merging the closest pair of threads — the two a
 * stitcher would least miss.
 *
 * Runs between the quantizer and everything else, on the same denoised cells the quantizer clustered.
 */

/** Hue bins around the circle. Twelve is coarse enough that two threads of one hue don't both count as coverage. */
export const HUE_BINS = 12;
/** A cell counts as coloured above this chroma; below it the hue of a near-neutral is noise. */
export const HUE_RESERVE_CHROMA_FLOOR = 0.02;
/** A bin earns a thread at this share of the stitched cells. */
export const HUE_RESERVE_MIN_SHARE = 0.001;
/** At most this many threads are reserved, so a chart keeps its tonal steps. */
export const HUE_RESERVE_MAX_THREADS = 6;

/**
 * A thread speaks for a hue only if it carries at least this share of the chroma the most colourful cell of that hue
 * has. A washed-out thread in the pink bin is not a pink thread — which is what made the cat's pink appear at 16 and
 * 24 colours and vanish again at 40, where the larger palette happened to put a near-neutral there. Comparing against
 * the bin's own best cell rather than a fixed chroma keeps a heavy, genuinely muted hue from claiming slots.
 */
export const HUE_RESERVE_COVERAGE_SHARE = 0.5;

/**
 * And it must be a colour in its own right. Below this chroma a thread reads as a neutral whatever bin it falls in,
 * so counting it as coverage is what let the cat's pink be "spoken for" by a thread nobody would call pink.
 */
export const HUE_RESERVE_COVERAGE_FLOOR = 0.03;

const FLOOR_SQUARED = HUE_RESERVE_CHROMA_FLOOR * HUE_RESERVE_CHROMA_FLOOR;
const COVERAGE_SHARE_SQUARED = HUE_RESERVE_COVERAGE_SHARE * HUE_RESERVE_COVERAGE_SHARE;
const COVERAGE_FLOOR_SQUARED = HUE_RESERVE_COVERAGE_FLOOR * HUE_RESERVE_COVERAGE_FLOOR;

export interface HueReserveResult {
  cellPaletteIndex: Uint8Array;
  palette: RGB[];
  /** How many threads were reserved; 0 means the inputs come back untouched. */
  reserved: number;
}

/** Squared chroma, so the floor and the "most colourful" comparisons never take a square root. */
function chromaSquared(l: number, a: number, b: number): number {
  return a * a + b * b;
}

/** Which of `HUE_BINS` a colour falls in. `Math.atan2` is bit-exact in both languages (fdlibm, as photo enhancement uses). */
export function hueBinOf(a: number, b: number): number {
  const turns = (Math.atan2(b, a) / (2 * Math.PI) + 1) % 1;
  return Math.min(HUE_BINS - 1, Math.floor(turns * HUE_BINS));
}

export function reserveHueThreads(
  cellOklab: Float64Array,
  cellPaletteIndex: Uint8Array,
  palette: RGB[],
  emptyMask?: Uint8Array | null
): HueReserveResult {
  const cellCount = cellPaletteIndex.length;
  if (palette.length < 2) return { cellPaletteIndex, palette, reserved: 0 };

  // What the cells hold: how much of each hue, and the most colourful cell carrying it.
  const binWeight = new Array<number>(HUE_BINS).fill(0);
  const binBest = new Array<number>(HUE_BINS).fill(-1);
  const binBestChroma = new Array<number>(HUE_BINS).fill(0);
  let stitched = 0;
  for (let i = 0; i < cellCount; i++) {
    if (emptyMask?.[i]) continue;
    stitched++;
    const l = cellOklab[i * 3];
    const a = cellOklab[i * 3 + 1];
    const b = cellOklab[i * 3 + 2];
    const chroma = chromaSquared(l, a, b);
    if (chroma < FLOOR_SQUARED) continue;
    const bin = hueBinOf(a, b);
    binWeight[bin]++;
    // Strictly greater, so the earliest cell wins a tie and the choice can't depend on iteration luck.
    if (chroma > binBestChroma[bin]) {
      binBestChroma[bin] = chroma;
      binBest[bin] = i;
    }
  }

  // Which of those the palette already speaks for, and how strongly.
  const oklab = palette.map(rgbToOklab);
  const binThreadChroma = new Array<number>(HUE_BINS).fill(0);
  for (const [, a, b] of oklab) {
    const chroma = chromaSquared(0, a, b);
    if (chroma < FLOOR_SQUARED) continue;
    const bin = hueBinOf(a, b);
    if (chroma > binThreadChroma[bin]) binThreadChroma[bin] = chroma;
  }
  const covered = binThreadChroma.map(
    (thread, bin) => thread >= COVERAGE_FLOOR_SQUARED && thread >= COVERAGE_SHARE_SQUARED * binBestChroma[bin]
  );

  const wanted: Array<{ bin: number; weight: number }> = [];
  for (let bin = 0; bin < HUE_BINS; bin++) {
    if (covered[bin] || binBest[bin] < 0) continue;
    if (binWeight[bin] < HUE_RESERVE_MIN_SHARE * stitched) continue;
    wanted.push({ bin, weight: binWeight[bin] });
  }
  // Heaviest hue first; a tie goes to the lower bin, so the order never depends on the sort's stability.
  wanted.sort((x, y) => (y.weight !== x.weight ? y.weight - x.weight : x.bin - y.bin));
  const taking = wanted.slice(0, HUE_RESERVE_MAX_THREADS);
  if (taking.length === 0) return { cellPaletteIndex, palette, reserved: 0 };

  const centroids: Oklab[] = oklab.map((c) => [c[0], c[1], c[2]]);
  const counts = new Array<number>(centroids.length).fill(0);
  for (let i = 0; i < cellCount; i++) {
    if (emptyMask?.[i]) continue;
    counts[cellPaletteIndex[i]]++;
  }

  for (const { bin } of taking) {
    // The slot is paid for by the closest pair in the palette: the two threads a stitcher would least miss.
    let pairI = 0;
    let pairJ = 1;
    let bestDist = Infinity;
    for (let i = 0; i < centroids.length; i++) {
      for (let j = i + 1; j < centroids.length; j++) {
        const d = oklabDistanceSquared(centroids[i], centroids[j]);
        if (d < bestDist) {
          bestDist = d;
          pairI = i;
          pairJ = j;
        }
      }
    }
    const loser = counts[pairI] <= counts[pairJ] ? pairI : pairJ;
    counts[loser === pairI ? pairJ : pairI] += counts[loser];
    const seed = binBest[bin];
    centroids[loser] = [cellOklab[seed * 3], cellOklab[seed * 3 + 1], cellOklab[seed * 3 + 2]];
    counts[loser] = 0;
  }

  // Each cell to its nearest thread, once. Re-converging Lloyd here loses every reserved hue: the seeded centroid
  // drifts back into the mass it was placed to escape (measured, `docs/reviews/2026-09-23-hue-reservation.md`).
  const assignment = new Uint8Array(cellCount);
  const used = new Array<number>(centroids.length).fill(0);
  for (let i = 0; i < cellCount; i++) {
    if (emptyMask?.[i]) {
      assignment[i] = EMPTY_CELL;
      continue;
    }
    const cell: Oklab = [cellOklab[i * 3], cellOklab[i * 3 + 1], cellOklab[i * 3 + 2]];
    let best = 0;
    let bestDist = Infinity;
    for (let c = 0; c < centroids.length; c++) {
      const d = oklabDistanceSquared(cell, centroids[c]);
      if (d < bestDist) {
        bestDist = d;
        best = c;
      }
    }
    assignment[i] = best;
    used[best]++;
  }

  const remap = new Int16Array(centroids.length).fill(-1);
  const nextPalette: RGB[] = [];
  centroids.forEach((centroid, c) => {
    if (used[c] === 0) return;
    remap[c] = nextPalette.length;
    nextPalette.push(oklabToRgb(centroid));
  });
  const nextIndex = new Uint8Array(cellCount);
  for (let i = 0; i < cellCount; i++) {
    nextIndex[i] = assignment[i] === EMPTY_CELL ? EMPTY_CELL : remap[assignment[i]];
  }

  return { cellPaletteIndex: nextIndex, palette: nextPalette, reserved: taking.length };
}

File: rust/cs-core/src/hue_reserve.rs
//! Port of `lib/pipeline/hue-reserve.ts`: Vivid, part two (G-062, D212).
//!
//! k-means allocates by squared error, so a colour covering half a percent of the chart loses to one more step in
//! the dominant ramp however visible it is. This gives each hue the cells hold, and the palette does not speak for,
//! a slot of its own, paid for by merging the closest pair of threads.

use crate::color::{oklab_distance_sq, oklab_to_rgb, rgb_to_oklab, Oklab, Rgb};
use crate::jsmath;

pub const HUE_BINS: usize = 12;
pub const HUE_RESERVE_CHROMA_FLOOR: f64 = 0.02;
pub const HUE_RESERVE_MIN_SHARE: f64 = 0.001;
pub const HUE_RESERVE_MAX_THREADS: usize = 6;
pub const HUE_RESERVE_COVERAGE_SHARE: f64 = 0.5;
pub const HUE_RESERVE_COVERAGE_FLOOR: f64 = 0.03;

const FLOOR_SQUARED: f64 = HUE_RESERVE_CHROMA_FLOOR * HUE_RESERVE_CHROMA_FLOOR;
const COVERAGE_SHARE_SQUARED: f64 = HUE_RESERVE_COVERAGE_SHARE * HUE_RESERVE_COVERAGE_SHARE;
const COVERAGE_FLOOR_SQUARED: f64 = HUE_RESERVE_COVERAGE_FLOOR * HUE_RESERVE_COVERAGE_FLOOR;

/// `Math.atan2` is bit-exact in both languages (fdlibm), as photo enhancement already relies on.
pub fn hue_bin_of(a: f64, b: f64) -> usize {
    let turns = (jsmath::atan2(b, a) / (2.0 * std::f64::consts::PI) + 1.0) % 1.0;
    let bin = (turns * HUE_BINS as f64).floor() as usize;
    bin.min(HUE_BINS - 1)
}

pub struct HueReserveResult {
    pub cell_palette_index: Vec<u8>,
    pub palette: Vec<Rgb>,
    pub reserved: usize,
}

pub fn reserve_hue_threads(
    cell_oklab: &[f64],
    cell_palette_index: &[u8],
    palette: &[Rgb],
    empty_mask: Option<&[u8]>,
) -> HueReserveResult {
    let cell_count = cell_palette_index.len();
    if palette.len() < 2 {
        return HueReserveResult {
            cell_palette_index: cell_palette_index.to_vec(),
            palette: palette.to_vec(),
            reserved: 0,
        };
    }

    let mut bin_weight = [0usize; HUE_BINS];
    let mut bin_best: [i64; HUE_BINS] = [-1; HUE_BINS];
    let mut bin_best_chroma = [0f64; HUE_BINS];
    let mut stitched = 0usize;
    for i in 0..cell_count {
        if empty_mask.is_some_and(|m| m[i] != 0) {
            continue;
        }
        stitched += 1;
        let a = cell_oklab[i * 3 + 1];
        let b = cell_oklab[i * 3 + 2];
        let chroma = a * a + b * b;
        if chroma < FLOOR_SQUARED {
            continue;
        }
        let bin = hue_bin_of(a, b);
        bin_weight[bin] += 1;
        if chroma > bin_best_chroma[bin] {
            bin_best_chroma[bin] = chroma;
            bin_best[bin] = i as i64;
        }
    }

    let oklab: Vec<Oklab> = palette.iter().map(|&c| rgb_to_oklab(c)).collect();
    let mut bin_thread_chroma = [0f64; HUE_BINS];
    for lab in &oklab {
        let chroma = lab[1] * lab[1] + lab[2] * lab[2];
        if chroma < FLOOR_SQUARED {
            continue;
        }
        let bin = hue_bin_of(lab[1], lab[2]);
        if chroma > bin_thread_chroma[bin] {
            bin_thread_chroma[bin] = chroma;
        }
    }

    let mut wanted: Vec<(usize, usize)> = Vec::new();
    for bin in 0..HUE_BINS {
        let covered = bin_thread_chroma[bin] >= COVERAGE_FLOOR_SQUARED
            && bin_thread_chroma[bin] >= COVERAGE_SHARE_SQUARED * bin_best_chroma[bin];
        if covered || bin_best[bin] < 0 {
            continue;
        }
        if (bin_weight[bin] as f64) < HUE_RESERVE_MIN_SHARE * stitched as f64 {
            continue;
        }
        wanted.push((bin, bin_weight[bin]));
    }
    // Heaviest hue first; a tie goes to the lower bin, so the order never depends on the sort's stability.
    wanted.sort_by(|x, y| y.1.cmp(&x.1).then(x.0.cmp(&y.0)));
    wanted.truncate(HUE_RESERVE_MAX_THREADS);
    if wanted.is_empty() {
        return HueReserveResult {
            cell_palette_index: cell_palette_index.to_vec(),
            palette: palette.to_vec(),
            reserved: 0,
        };
    }

    let mut centroids = oklab.clone();
    let mut counts = vec![0usize; centroids.len()];
    for i in 0..cell_count {
        if empty_mask.is_some_and(|m| m[i] != 0) {
            continue;
        }
        counts[cell_palette_index[i] as usize] += 1;
    }

    for &(bin, _) in &wanted {
        let mut pair_i = 0usize;
        let mut pair_j = 1usize;
        let mut best_dist = f64::INFINITY;
        for i in 0..centroids.len() {
            for j in i + 1..centroids.len() {
                let d = oklab_distance_sq(&centroids[i], &centroids[j]);
                if d < best_dist {
                    best_dist = d;
                    pair_i = i;
                    pair_j = j;
                }
            }
        }
        let loser = if counts[pair_i] <= counts[pair_j] { pair_i } else { pair_j };
        let winner = if loser == pair_i { pair_j } else { pair_i };
        counts[winner] += counts[loser];
        let seed = bin_best[bin] as usize;
        centroids[loser] = [
            cell_oklab[seed * 3],
            cell_oklab[seed * 3 + 1],
            cell_oklab[seed * 3 + 2],
        ];
        counts[loser] = 0;
    }

    // Each cell to its nearest thread, once: re-converging Lloyd here loses every reserved hue.
    let mut assignment = vec![0u8; cell_count];
    let mut used = vec![0usize; centroids.len()];
    for i in 0..cell_count {
        if empty_mask.is_some_and(|m| m[i] != 0) {
            assignment[i] = crate::EMPTY_CELL;
            continue;
        }
        let cell: Oklab = [cell_oklab[i * 3], cell_oklab[i * 3 + 1], cell_oklab[i * 3 + 2]];
        let mut best = 0usize;
        let mut best_dist = f64::INFINITY;
        for (c, centroid) in centroids.iter().enumerate() {
            let d = oklab_distance_sq(&cell, centroid);
            if d < best_dist {
                best_dist = d;
                best = c;
            }
        }
        assignment[i] = best as u8;
        used[best] += 1;
    }

    let mut remap = vec![-1i32; centroids.len()];
    let mut next_palette: Vec<Rgb> = Vec::new();
    for (c, centroid) in centroids.iter().enumerate() {
        if used[c] == 0 {
            continue;
        }
        remap[c] = next_palette.len() as i32;
        next_palette.push(oklab_to_rgb(*centroid));
    }
    let next_index: Vec<u8> = assignment
        .iter()
        .map(|&a| {
            if a == crate::EMPTY_CELL {
                crate::EMPTY_CELL
            } else {
                remap[a as usize] as u8
            }
        })
        .collect();

    HueReserveResult {
        cell_palette_index: next_index,
        palette: next_palette,
        reserved: wanted.len(),
    }
}
